File: python/alternating_case.py
# 8 kyu
# altERnaTIng cAsE <=> ALTerNAtiNG CaSe
#
# Define String.prototype.toAlternatingCase (or a similar function/method
# such as to_alternating_case/toAlternatingCase/ToAlternatingCase in your selected language;
# see the initial solution for details) such that each lowercase letter becomes uppercase and each uppercase letter becomes lowercase.

import codewars_test as test


def to_alternating_case(string) -> str:
    return "".join(i.upper() if i.islower() else i.lower() for i in string)


# clever
# return string.swapcase()


@test.describe("Basic tests")
def test_bacics():
    @test.it("should work for fixed tests (provided in the description)")
    def test_fixed():
        test.assert_equals(to_alternating_case("hello world"), "HELLO WORLD")
        test.assert_equals(to_alternating_case("HELLO WORLD"), "hello world")
        test.assert_equals(to_alternating_case("hello WORLD"), "HELLO world")
        test.assert_equals(to_alternating_case("HeLLo WoRLD"), "hEllO wOrld")
        test.assert_equals(to_alternating_case("12345"), "12345")
        test.assert_equals(to_alternating_case("1a2b3c4d5e"), "1A2B3C4D5E")
        test.assert_equals(
            to_alternating_case("String.prototype.toAlternatingCase"),
            "sTRING.PROTOTYPE.TOaLTERNATINGcASE",
        )
        test.assert_equals(
            to_alternating_case(to_alternating_case("Hello World")), "Hello World"
        )

    @test.it("should work for the title of this Kata")
    def test_title():
        title = "altERnaTIng cAsE"
        title = to_alternating_case(title)
        test.assert_equals(title, "ALTerNAtiNG CaSe")
        title = to_alternating_case(title)
        test.assert_equals(title, "altERnaTIng cAsE")
        title = to_alternating_case(title)
        test.assert_equals(title, "ALTerNAtiNG CaSe")
        title = to_alternating_case(title)
        test.assert_equals(title, "altERnaTIng cAsE")
        title = "altERnaTIng cAsE <=> ALTerNAtiNG CaSe"
        title = to_alternating_case(title)
        test.assert_equals(title, "ALTerNAtiNG CaSe <=> altERnaTIng cAsE")
        title = to_alternating_case(title)
        test.assert_equals(title, "altERnaTIng cAsE <=> ALTerNAtiNG CaSe")
        title = to_alternating_case(title)
        test.assert_equals(title, "ALTerNAtiNG CaSe <=> altERnaTIng cAsE")
        title = to_alternating_case(title)
        test.assert_equals(title, "altERnaTIng cAsE <=> ALTerNAtiNG CaSe")

File: rust/src/alternating_case.rs
/*
    8 kyu
    altERnaTIng cAsE <=> ALTerNAtiNG CaSe
    https://www.codewars.com/kata/5899904a3e4df800b1000132/train/rust
    Define a function to_alternating_case such that each lowercase letter
    becomes uppercase and each uppercase letter becomes lowercase.
*/

fn to_alternating_case(s: &str) -> String {
    let mut result = String::new();
    s.chars()
        .flat_map(|c: char| {
            if c.is_uppercase() {
                c.to_lowercase()
            } else {
                c.to_uppercase()
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn example_tests() {
        assert_eq!("HELLO WORLD", to_alternating_case("hello world"));
        assert_eq!("hello world", to_alternating_case("HELLO WORLD"));
        assert_eq!("HELLO world", to_alternating_case("hello WORLD"));
        assert_eq!("hEllO wOrld", to_alternating_case("HeLLo WoRLD"));
        assert_eq!(
            "Hello World",
            to_alternating_case(&to_alternating_case("Hello World")[..])
        );
        assert_eq!("12345", to_alternating_case("12345"));
        assert_eq!("1A2B3C4D5E", to_alternating_case("1a2b3c4d5e"));
        assert_eq!(
            "sTRING.tOaLTERNATINGcASE",
            to_alternating_case("String.ToAlternatingCase")
        );
    }
}
